File: round_robin.py
def round_robin(arrival_time, burst_time, quantum, number_of_processes):

  pq = []
  for i in range(number_of_processes):
    pq.append([arrival_time[i], burst_time[i], i])

  completion_times = []
  current_time = 0
  while len(pq) > 0:
    process = pq.pop(0)
    if arrival_time[process[2]] <= current_time:
      current_time += min(process[1], quantum)
      completion_times.append(current_time)
      process[1] -= min(process[1], quantum)
      if process[1] > 0:
        pq.append(process)
    else:
      current_time = arrival_time[process[2]] + min(process[1], quantum)
      completion_times.append(current_time)
      process[1] -= min(process[1], quantum)
      if process[1] > 0:
        pq.append(process)

  turn_around_times = []
  waiting_times = []
  for i in range(number_of_processes):
    turn_around_times.append(completion_times[i] - arrival_time[i])
    waiting_times.append(turn_around_times[i] - burst_time[i])

  average_waiting_time = sum(waiting_times) / number_of_processes
  average_tat= sum(turn_around_times) / number_of_processes

  return completion_times, turn_around_times, waiting_times, average_waiting_time,average_tat


if __name__ == "__main__":
  arrival_time = [0, 1, 2, 3]
  burst_time = [4, 5, 2, 1]
  number_of_processes = 4
  quantum = 2

  completion_times, turn_around_times, waiting_times, average_waiting_time,average_tat = round_robin(arrival_time, burst_time, quantum, number_of_processes)
  print("Completion times:", completion_times)
  print("Turn around times:", turn_around_times)
  print("Waiting times:", waiting_times)
  print("Average waiting time:", average_waiting_time)
  print("Average turn around time:", average_tat)

File: round_robin.c
#include <stdio.h>

#define MAX_PROCESSES 50

void calculateTimes(int n, int burstTime[], int timeQuantum, int completionTime[], int waitingTime[], int turnaroundTime[]) {
    int remainingTime[MAX_PROCESSES];
    int currentTime = 0;
    int completed = 0;

    for (int i = 0; i < n; i++) {
        remainingTime[i] = burstTime[i];
    }

    while (completed < n) {
        for (int i = 0; i < n; i++) {
            if (remainingTime[i] > 0) {
                if (remainingTime[i] <= timeQuantum) {
                    currentTime += remainingTime[i];
                    completionTime[i] = currentTime;
                    remainingTime[i] = 0;
                    completed++;
                } else {
                    currentTime += timeQuantum;
                    remainingTime[i] -= timeQuantum;
                }
            }
        }
    }

    for (int i = 0; i < n; i++) {
        turnaroundTime[i] = completionTime[i];
        waitingTime[i] = turnaroundTime[i] - burstTime[i];
    }
}

int main() {
    int n, timeQuantum;
    printf("Enter the number of processes: ");
    scanf("%d", &n);

    int burstTime[MAX_PROCESSES];
    int completionTime[MAX_PROCESSES];
    int waitingTime[MAX_PROCESSES];
    int turnaroundTime[MAX_PROCESSES];

    printf("Enter time quantum: ");
    scanf("%d", &timeQuantum);

    printf("Enter burst time for each process:\n");
    for (int i = 0; i < n; i++) {
        printf("Process %d: ", i + 1);
        scanf("%d", &burstTime[i]);
    }

    calculateTimes(n, burstTime, timeQuantum, completionTime, waitingTime, turnaroundTime);

    printf("\nProcess\t Burst Time\t Completion Time\t Waiting Time\t Turnaround Time\n");
    for (int i = 0; i < n; i++) {
        printf("%d\t %d\t\t %d\t\t %d\t\t %d\n",
               i + 1, burstTime[i], completionTime[i], waitingTime[i], turnaroundTime[i]);
    }

    // Calculate and display average waiting time and average turnaround time
    float avgWaitingTime = 0, avgTurnaroundTime = 0;
    for (int i = 0; i < n; i++) {
        avgWaitingTime += waitingTime[i];
        avgTurnaroundTime += turnaroundTime[i];
    }
    avgWaitingTime /= n;
    avgTurnaroundTime /= n;

    printf("\nAverage Waiting Time: %.2f\n", avgWaitingTime);
    printf("Average Turnaround Time: %.2f\n", avgTurnaroundTime);

    return 0;
}
